src/lib/logging/binary_formatter.cpp

```diff
@@ -0,0 +1,218 @@
+/*
+ *  Formatter that creates binary log entries.
+ * 
+ *  The log entries have following format:
+ *         
+ *         Value                : Number of Bytes                         : Description
+ * 
+ *     Commit Entries:
+ *       - log entry type ('c') : sizeof(char)
+ *       - transaction_id       : sizeof(TransactionID)
+ * 
+ *     Value Entries:
+ *       - log entry type ('v') : sizeof(char)
+ *       - transaction_id       : sizeof(TransactionID)
+ *       - table_name           : table_name.size() + 1                   : string terminated with \0
+ *       - row_id               : sizeof(ChunkID) + sizeof(ChunkOffset)
+ *       - NULL bitmap          : ceil(values.size() / 8.0)               : Bitmap indicating NullValues with 1
+ *       - value                : length(value)
+ *       - any optional values
+ * 
+ *     Invalidation Entries:
+ *       - log entry type ('i') : sizeof(char)
+ *       - transaction_id       : sizeof(TransactionID)
+ *       - table_name           : table_name.size() + 1                   : string terminated with \0
+ *       - row_id               : sizeof(ChunkID) + sizeof(ChunkOffset) 
+ *
+ *     Load Table Entries:
+ *       - log entry type ('l') : sizeof(char)
+ *       - file_path            : file_path.size() + 1                    : string terminated with \0
+ *       - table_name           : table_name.size() + 1                   : string terminated with \0
+ * 
+ * 
+ *  Potential Improvements:
+ *    1.  For each log entry a vector<char> is allocated to create that entry and then copy it into the buffer.
+ *        Maybe allocate a big memory block once.
+ *    2.  The entry vector gets resized for each value. Maybe .reserve() beforehand or calculate the number of bytes for
+ *        all values by iterating over them before putting them into the entry.
+ *        Then the vector needs to be resized just once.
+ */
+
+#include "binary_formatter.hpp"
+
+#include "all_type_variant.hpp"
+#include "binary_recoverer.hpp"
+#include "types.hpp"
+
+namespace opossum {
+
+// LogEntry is used to create a single log entry.
+// It keeps track of the cursor position while writing values into a buffer.
+class LogEntry {
+ private:
+  friend class EntryWriter;
+  uint32_t cursor{0};
+  std::vector<char> data;
+
+  explicit LogEntry(uint32_t count) { data.resize(count); }
+
+  void resize(size_t size) {
```

Comment:
For the current design the vector<char> needs to be resized. Since I write to it via &data[cursor] the vector is supposed to be of the needed size. (Reserve updates the capacity but not the size).